Updated the typing spec to use modern (PEP 695) generics syntax

erictraut opened the pull request:
This PR attempts to update all places in the spec where the old-style `TypeVar`, `ParamSpec` and `TypeVarTuple` constructors are used in code samples, with the exception of samples that are intended to explain the older syntax. This should improve readability.

Since it doesn't make any substantive changes to the spec, I don't think this requires approval from the full typing council, but it would be good to have a thorough review.

AlexWaygood commented on docs/spec/protocol.rst:
```diff
@@ -266,25 +266,25 @@ Generic protocols are important. For example, ``SupportsAbs``, ``Iterable``
+``Generic[T, S, ...]`` or to mix it with the new ``class Iterable[T]`` form::
 
-  class Iterable(Protocol[T], Generic[T]):   # INVALID
+  class Iterable[T](Protocol, Generic[T]):   # INVALID
```
I agree that this modified example demonstrates something a type checker should reject, but I don't think it demonstrates the same thing that it used to (and it seems to me that it makes the example immediately below it redundant). I'd revert this change

erictraut commented:
Thanks for the review @AlexWaygood. I think I've addressed your feedback.

I'm going to leave this unmerged a bit longer to give others a chance to review if they'd like.

JelleZijlstra commented on docs/spec/literal.rst:
```diff
@@ -733,18 +733,16 @@ Conditional statements and expressions work as expected:
 
-Interaction with TypeVars and Generics
-""""""""""""""""""""""""""""""""""""""
+Interaction with Type Variables and Generics
```
Out of scope for this PR but I feel this whole section is basically redundant and could be deleted. LiteralString is a type, so it can be used in any place that an arbitrary type can be used. Some of the Literal-related text above is similarly redundant.

JelleZijlstra commented on docs/spec/protocol.rst:
```diff
@@ -266,25 +266,26 @@ Generic protocols are important. For example, ``SupportsAbs``, ``Iterable``
-the declared variance. Examples::
+When using the generics syntax introduced in Python 3.12, the variance of
+type variables is inferred. When using the pre-3.12 generics syntax, variance
+must be specified. Type checkers will warn if the declared variance does not
```
```suggestion
must be specified (unless ``infer_variance=True`` is used). Type checkers will warn if the declared variance does not
```

JelleZijlstra reviewed (approved):
Thank you!
